src/DefaultBuilder/src/WebApplicationBuilder.cs

```diff
@@ -154,9 +145,76 @@ internal WebApplicationBuilder(WebApplicationOptions options, bool slim, Action<
                 options.SuppressEnvironmentConfiguration = true;
             });
 
+        _genericWebHostServiceDescriptor = InitializeHosting(bootstrapHostBuilder);
+    }
+
+    internal WebApplicationBuilder(WebApplicationOptions options, bool slim, bool empty, Action<IHostBuilder>? configureDefaults = null)
+    {
+        Debug.Assert(!slim, "should only be called with slim: false");
+        Debug.Assert(empty, "should only be called with empty: true");
+
+        var configuration = new ConfigurationManager();
+
+        // empty builder should still default the ContentRoot as usual. This is the expected behavior for all WebApplicationBuilders.
+        SetDefaultContentRoot(options, configuration);
+
+        _hostApplicationBuilder = Microsoft.Extensions.Hosting.Host.CreateEmptyApplicationBuilder(new HostApplicationBuilderSettings
+        {
+            Args = options.Args,
+            ApplicationName = options.ApplicationName,
+            EnvironmentName = options.EnvironmentName,
+            ContentRootPath = options.ContentRootPath,
+            Configuration = configuration,
+        });
+
+        // Set WebRootPath if necessary
+        if (options.WebRootPath is not null)
+        {
+            Configuration.AddInMemoryCollection(new[]
+            {
+                new KeyValuePair<string, string?>(WebHostDefaults.WebRootKey, options.WebRootPath),
+            });
+        }
+
+        // Run methods to configure web host defaults early to populate services
+        var bootstrapHostBuilder = new BootstrapHostBuilder(_hostApplicationBuilder);
+
+        // This is for testing purposes
+        configureDefaults?.Invoke(bootstrapHostBuilder);
+
+        bootstrapHostBuilder.ConfigureSlimWebHost(
+            webHostBuilder =>
+            {
+                // Note this doesn't configure any WebHost server - Kestrel or otherwise.
+                // It also doesn't register Routing, HostFiltering, or ForwardedHeaders.
+                // It is "empty" and up to the caller to configure these services.
+
+                // Runs inline.
+                webHostBuilder.Configure((context, app) => ConfigureApplication(context, app, allowDeveloperExceptionPage: false));
+
+                webHostBuilder.UseSetting(WebHostDefaults.ApplicationKey, _hostApplicationBuilder.Environment.ApplicationName ?? "");
+
+                // NOTE: There is no way to add Configuration before this gets called, so it is unnecessary
```

Comment:
This is not true given that `options.Args` can be passed in, and those should be respected. This feels like an important escape hatch in case someone using the empty builder wants to modify the hosting startup assemblies.

Reply:
Fixed and refactored.